arroyo-compiler-service/src/lib.rs

```diff
@@ -0,0 +1,226 @@
+
+    let addr = format!("0.0.0.0:{}", grpc).parse().unwrap();
+
+    info!("Starting compiler service at {}", addr);
+    info!(
+        "artifacts will be written to {}",
+        service.storage.canonical_url()
+    );
+
+    let last_used = service.last_used.clone();
+
+    if let Ok(idle_time) =
+        std::env::var("IDLE_SHUTDOWN_MS").map(|t| Duration::from_millis(u64::from_str(&t).unwrap()))
+    {
+        tokio::spawn(async move {
+            loop {
+                if from_millis(last_used.load(Ordering::Relaxed))
+                    .elapsed()
+                    .unwrap()
+                    > idle_time
+                {
+                    println!("Idle time exceeded, shutting down");
+                    exit(0);
+                }
+                tokio::time::sleep(Duration::from_secs(1)).await;
+            }
+        });
+    }
+
+    shutdown.spawn_task(
+        "grpc",
+        Server::builder()
+            .max_frame_size(Some((1 << 24) - 1)) // 16MB
+            .add_service(CompilerGrpcServer::new(service))
+            .serve(addr),
+    );
+
+    let _ = shutdown.wait_for_shutdown(Duration::from_secs(30)).await;
+}
+
+pub struct CompileService {
+    build_dir: PathBuf,
+    lock: Arc<Mutex<()>>,
+    last_used: Arc<AtomicU64>,
+    storage: StorageProvider,
+}
+
+impl CompileService {
+    pub async fn new() -> Self {
+        let build_dir = std::env::var("BUILD_DIR").unwrap_or("build_dir".to_string());
+
+        let artifact_url = std::env::var(ARTIFACT_URL_ENV)
+            .unwrap_or_else(|_| panic!("{} must be set", ARTIFACT_URL_ENV));
+
+        let storage = StorageProvider::for_url(&artifact_url)
+            .await
+            .expect("unable to construct storage provider");
+
+        let last_used = Arc::new(AtomicU64::new(to_millis(SystemTime::now())));
+
+        CompileService {
+            build_dir: PathBuf::from_str(&build_dir).unwrap(),
+            lock: Arc::new(Mutex::new(())),
+            last_used: last_used.clone(),
+            storage,
+        }
+    }
+
+    async fn write_udf_crate(&self, udf_crate: UdfCrate) -> anyhow::Result<()> {
+        let udf_build_dir = self.build_dir.join("udfs_dir/udf");
+
+        tokio::fs::create_dir_all(&udf_build_dir.join("src")).await?;
+        tokio::fs::write(udf_build_dir.join("src/lib.rs"), &udf_crate.definition).await?;
+        tokio::fs::write(udf_build_dir.join("Cargo.toml"), &udf_crate.cargo_toml).await?;
+
+        let udf_wrapper_dir = self.build_dir.join("udfs_dir").join("udf_wrapper");
+        tokio::fs::create_dir_all(&udf_wrapper_dir.join("src")).await?;
+        tokio::fs::write(udf_wrapper_dir.join("src/lib.rs"), &udf_crate.lib_rs).await?;
+
+        let udf_wrapper_cargo_toml = format!(
+            r#"
+[package]
+name = "udf_wrapper"
+version = "0.1.0"
+edition = "2021"
+
+
+[dependencies]
+udf = {{ path = "../udf" }}
+arrow = {{ version = "50.0.0", features = ["ffi"] }}
```

Comment:
This could probably be smaller, I think just arrow-array and arrow-data. Building with --release for me went from 36 seconds to 6.

Reply:
Unfortunately the functions `from_ffi` and `to_ffi` are only exported by the `arrow` crate.